components/style/atomic_refcell.rs

```diff
@@ -2,121 +2,358 @@
 
-pub trait Map<'a, Base, Curr> {
-    fn map<New, F>(self, f: F) -> MappedAtomicRef<'a, Base, New>
-        where F: FnOnce(&Curr) -> &New;
+impl<T: ?Sized + PartialOrd> PartialOrd for AtomicRefCell<T> {
+    #[inline]
+    fn partial_cmp(&self, other: &AtomicRefCell<T>) -> Option<cmp::Ordering> {
+        self.borrow().partial_cmp(&*other.borrow())
+    }
+
+    #[inline]
+    fn lt(&self, other: &AtomicRefCell<T>) -> bool {
+        *self.borrow() < *other.borrow()
+    }
+
+    #[inline]
+    fn le(&self, other: &AtomicRefCell<T>) -> bool {
+        *self.borrow() <= *other.borrow()
+    }
+
+    #[inline]
+    fn gt(&self, other: &AtomicRefCell<T>) -> bool {
+        *self.borrow() > *other.borrow()
+    }
+
+    #[inline]
+    fn ge(&self, other: &AtomicRefCell<T>) -> bool {
+        *self.borrow() >= *other.borrow()
+    }
 }
 
-impl<'a, Base> Map<'a, Base, Base> for AtomicRef<'a, Base> {
-    fn map<New, F>(self, f: F) -> MappedAtomicRef<'a, Base, New>
-        where F: FnOnce(&Base) -> &New
-    {
-        OwningRef::new(self).map(f)
+impl<T: ?Sized + Ord> Ord for AtomicRefCell<T> {
+    #[inline]
+    fn cmp(&self, other: &AtomicRefCell<T>) -> cmp::Ordering {
+        self.borrow().cmp(&*other.borrow())
+    }
+}
+
+impl<T> From<T> for AtomicRefCell<T> {
+    fn from(t: T) -> AtomicRefCell<T> {
+        AtomicRefCell::new(t)
+    }
+}
+
+impl<'b> Clone for AtomicBorrowRef<'b> {
+    #[inline]
+    fn clone(&self) -> AtomicBorrowRef<'b> {
+        AtomicBorrowRef::new(self.borrow)
     }
 }
 
-impl<'a, Base, Curr> Map<'a, Base, Curr> for MappedAtomicRef<'a, Base, Curr> {
-    fn map<New, F>(self, f: F) -> MappedAtomicRef<'a, Base, New>
-        where F: FnOnce(&Curr) -> &New
+/// A wrapper type for an immutably borrowed value from an `AtomicRefCell<T>`.
+pub struct AtomicRef<'b, T: ?Sized + 'b> {
+    value: &'b T,
+    borrow: AtomicBorrowRef<'b>,
+}
+
+
+impl<'b, T: ?Sized> Deref for AtomicRef<'b, T> {
+    type Target = T;
+
+    #[inline]
+    fn deref(&self) -> &T {
+        self.value
+    }
+}
+
+impl<'b, T: ?Sized> AtomicRef<'b, T> {
+    /// Copies an `AtomicRef`.
+    #[inline]
+    pub fn clone(orig: &AtomicRef<'b, T>) -> AtomicRef<'b, T> {
+        AtomicRef {
+            value: orig.value,
+            borrow: orig.borrow.clone(),
+        }
+    }
+
+    /// Make a new `AtomicRef` for a component of the borrowed data.
+    #[inline]
+    pub fn map<U: ?Sized, F>(orig: AtomicRef<'b, T>, f: F) -> AtomicRef<'b, U>
+        where F: FnOnce(&T) -> &U
     {
-        self.map(f)
+        AtomicRef {
+            value: f(orig.value),
+            borrow: orig.borrow,
+        }
     }
 }
 
-impl<'a, Base> AtomicRef<'a, Base> {
-    pub fn map<Curr, New, M, F>(orig: M, f: F) -> MappedAtomicRef<'a, Base, New>
-        where F: FnOnce(&Curr) -> &New, M: Map<'a, Base, Curr>
+impl<'b, T: ?Sized> AtomicRefMut<'b, T> {
+    /// Make a new `AtomicRefMut` for a component of the borrowed data, e.g. an enum
+    /// variant.
```

Comment:
I think an enum variant would want an `and_then` or `filter_map` function; `map` seems more useful for struct fields.